### PR TITLE
8342486: Implement JEP draft: Structured Concurrency (Fourth Preview)

### PR DESCRIPTION
**Pull request by AlanBateman**
Updates for JEP draft: Structured Concurrency (Fourth Preview).

As always, any questions/comments on the API should be brought to loom-dev.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires CSR request [JDK-8342487](https://bugs.openjdk.org/browse/JDK-8342487) to be approved

### Issues
 * [JDK-8342486](https://bugs.openjdk.org/browse/JDK-8342486): Implement JEP draft: Structured Concurrency (Fourth Preview) (**Enhancement** - P3)
 * [JDK-8342487](https://bugs.openjdk.org/browse/JDK-8342487): Implement JEP draft: Structured Concurrency (Fourth Preview) (**CSR**)

### Contributors
 * Alan Bateman `<alanb@openjdk.org>`
 * Viktor Klang `<vklang@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21904/head:pull/21904` \
`$ git checkout pull/21904`

Update a local copy of the PR: \
`$ git checkout pull/21904` \
`$ git pull https://git.openjdk.org/jdk.git pull/21904/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21904`

View PR using the GUI difftool: \
`$ git pr show -t 21904`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21904.diff">https://git.openjdk.org/jdk/pull/21904.diff</a>

</details>

### CODE REVIEW

**Comment by liach** [comment]
API questions:
1. Do we guarantee any ordering for each `onFork`, `onComplete`, and `result`, like "actions in an `onComplete` that returns true happens-before the execution of `result`"?
2. For `Joiner::onFork` and `onComplete`, should we note what happens if users throw exceptions in these methods? The default implementation is already throwing exceptions, but readers don't know what the implications are.
3. About config: did you find that in practice, config reuse is so rare that allowing users to keep config objects as a static final field (like ClassFile API's `ClassFile` config) is infeasible, so you opted for a function to update config instead?

File: src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java [hunk 1]
```diff
@@ -24,317 +24,347 @@
...
- * A basic API for <em>structured concurrency</em>. {@code StructuredTaskScope} supports
- * cases where a task splits into several concurrent subtasks, and where the subtasks must
- * complete before the main task continues. A {@code StructuredTaskScope} can be used to
- * ensure that the lifetime of a concurrent operation is confined by a <em>syntax block</em>,
- * just like that of a sequential operation in structured programming.
+ * An API for <em>structured concurrency</em>. {@code StructuredTaskScope} supports cases
+ * where execution of a <em>task</em> (a unit of work) splits into several concurrent
+ * subtasks, and where the subtasks must complete before the task continues. A {@code
+ * StructuredTaskScope} can be used to ensure that the lifetime of a concurrent operation
+ * is confined by a <em>syntax block</em>, just like that of a sequential operation in
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java:39]:
```suggestion
 * is confined by a <em>syntax block</em>, similar to that of a sequential operation in
```

File: src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java [hunk 2]
```diff
@@ -24,317 +24,347 @@
...
+ * The scope is cancelled if any of the subtasks fail and {@code join} throws {@code
+ * FailedException} with the exception from the failed subtask as the cause. Other {@code
+ * Joiner} implementations may cancel the scope for other reasons.
+ *
+ * <p> To allow for cancellation, subtasks must be coded so that they finish as soon as
+ * possible when interrupted. Subtasks that do not respond to interrupt, e.g. block on
+ * methods that are not interruptible, may delay the closing of a scope indefinitely. The
+ * {@link #close() close} method always waits for threads executing subtasks to finish,
+ * even if the scope is cancelled, so execution cannot continue beyond the {@code close}
+ * method until the interrupted threads finish.
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java:122]: I'm thinking we might want to make this point stand out a bit more, as it is going to be important information for troubleshooting.

File: src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java [hunk 3]
```diff
@@ -24,317 +24,347 @@
...
- * subtask to complete successfully, cancelling the other by way of shutting down the task
- * scope. The main task waits in {@code join} until either subtask completes with a result
- * or both subtasks fail. It invokes {@link ShutdownOnSuccess#result(Function)
- * result(Function)} method to get the captured result. If both subtasks fail then this
- * method throws a {@code WebApplicationException} with the exception from one of the
- * subtasks as the cause.
+ * <p> Many of the details for how exceptions are handled will depend on usage. In some
+ * cases it may be useful to add a {@code catch} block to the {@code try}-with-resources
+ * statement to catch {@code FailedException}. The exception handling may use {@code
+ * instanceof} with pattern matching to handle specific causes.
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java:175]: 👍

File: src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java [hunk 4]
```diff
@@ -344,977 +374,691 @@ public sealed interface Subtask<T> extends Supplier<T> permits SubtaskImpl {
...
+     * different scopes or re-used after a task is closed.
      *
-     * @param name the name of the task scope, can be null
-     * @param factory the thread factory
+     * <p> Designing a {@code Joiner} should take into account the code at the use-site
+     * where the results from the {@link StructuredTaskScope#join() join} method are
+     * processed. It should be clear what the {@code Joiner} does vs. the application
+     * code at the use-site. In general, the {@code Joiner} implementation is not the
+     * place to code "business logic". A {@code Joiner} should be designed to be as
+     * general purpose as possible.
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java:498]: 👍

File: src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java [hunk 5]
```diff
@@ -344,977 +374,691 @@ public sealed interface Subtask<T> extends Supplier<T> permits SubtaskImpl {
...
+         * complete without any cancellation. This is similar to {@link #awaitAll()}
+         * except that it yields a stream of the completed subtasks.
+         * {@snippet lang=java :
+         *    <T> List<Subtask<T>> invokeAll(Collection<Callable<T>> tasks) throws InterruptedException {
+         *        try (var scope = StructuredTaskScope.open(Joiner.<T>allUntil(_ -> false))) {
+         *            tasks.forEach(scope::fork);
+         *            return scope.join().toList();
+         *        }
+         *    }
+         * }
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java:714]: 👍

File: src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java [hunk 6]
```diff
@@ -344,977 +374,691 @@ public sealed interface Subtask<T> extends Supplier<T> permits SubtaskImpl {
...
-     * Throws IllegalStateException if the current thread is the owner, and the owner did
-     * not join after forking a subtask in the given fork round.
+     * Opens a new {@code StructuredTaskScope} to use the given {@code Joiner} object and
+     * with configuration that is the result of applying the given function to the
+     * <a href="#DefaultConfiguration">default configuration</a>.
+     *
+     * <p> The {@code configFunction} is called with the default configuration and returns
+     * the configuration for the new scope. The function may, for example, set the
+     * {@linkplain Config#withThreadFactory(ThreadFactory) ThreadFactory} or set
+     * a {@linkplain Config#withTimeout(Duration) timeout}.
```
Comment (viktorklang-ora) [inline src/java.base/share/classes/java/util/concurrent/StructuredTaskScope.java:835]: Do we need to clarify that the configFunction must return a non-null value?

**Comment by AlanBateman** [comment]
This was just a test PR, I didn't spot that he had comments.

> 1. Do we guarantee any ordering for each `onFork`, `onComplete`, and `result`, like "actions in an `onComplete` that returns true happens-before the execution of `result`"?

Viktor and I have discussed expanding the "Memory consistency effects" section but didn't decide on adding further statements about HB. It's a good topic though, and applies equally to the existing API, not just this update.


> 2. For `Joiner::onFork` and `onComplete`, should we note what happens if users throw exceptions in these methods? The default implementation is already throwing exceptions, but readers don't know what the implications are.

fork(Callable) does specify that it propagates any exception/error from onFork, but maybe you are asking if it should be repeated in the Joiner interface? Same thing for onComplete.

> 3. About config: did you find that in practice, config reuse is so rare that allowing users to keep config objects as a static final field (like ClassFile API's `ClassFile` config) is infeasible, so you opted for a function to update config instead?

The configuration includes timeouts and name that are very much per instance. We prototyped static configuration and builders, finally settled on the function to produce the configuration. This pattern isn't used much but it is useful for cases like this.

**Pull request closed without merging**